### PR TITLE
feat(url): support AWS S3 signed images in trace/observation preview

### PR DESCRIPTION
**Pull request by marliessophie**
> [!IMPORTANT]
> Adds AWS S3 signed URL validation and updates logging in `utilities.ts`.
> 
>   - **Behavior**:
>     - Adds `isValidPresignedS3Url()` in `utilities.ts` for AWS S3 signed URL validation.
>     - Updates `isValidImageUrl()` to incorporate `isValidPresignedS3Url()` for S3 URL handling.
>     - Replaces `console.log` with `logger.info` for logging in `utilities.ts`.
>   - **Validation**:
>     - Introduces `s3UrlSchema` for AWS S3 URL structure validation.
>     - Handles HEAD and GET requests for S3 URL validation, checking for 200, 206, and 403 status codes.
> 
> <sup>This description was created by </sup>[Ellipsis](https://www.ellipsis.dev?ref=langfuse%2Flangfuse&utm_source=github&utm_medium=referral)<sup> for 111b0a8676a244c2d6d8b54c7acd38bde6860864. It will automatically update as commits are pushed.</sup>

### CODE REVIEW

File: web/src/server/api/routers/utilities.ts [hunk 1]
```diff
@@ -81,17 +82,102 @@ const isValidAndSecureUrl = async (urlString: string): Promise<boolean> => {
...
+    "X-Amz-Algorithm": z.string().optional(),
+    "X-Amz-Credential": z.string().optional(),
+    "X-Amz-Date": z.string().optional(),
+    "X-Amz-Expires": z.string().optional(),
+    "X-Amz-SignedHeaders": z.string().optional(),
+    "X-Amz-Signature": z.string(),
+    "X-Amz-Security-Token": z.string().optional(),
+    "x-id": z.string().optional(), // Optional ID parameter, not documented in AWS docs as it is SDK specific, https://github.com/aws/aws-sdk-go-v2/blob/04e7aca073a0a7ed479aa37cad88a1cf58a979a1/service/s3/internal/customizations/presign_test.go#L35-L41
+  }),
+});
```
Comment (Steffen911) [inline web/src/server/api/routers/utilities.ts:108]: Do we also have to consider presigned URLs for Azure, GCP, ... here? Those may be relevant for self-hosters.

File: web/src/server/api/routers/utilities.ts [hunk 2]
```diff
@@ -81,17 +82,102 @@ const isValidAndSecureUrl = async (urlString: string): Promise<boolean> => {
...
+    }
     return false;
   }
 };
 
 const isValidImageUrl = async (url: string): Promise<boolean> => {
   try {
+    if (await isValidPresignedS3Url(url)) {
+      return true;
+    }
```
Comment (Steffen911) [inline web/src/server/api/routers/utilities.ts:174]: Do we have to validate whether the presigned URL actually points to an image (which may make sense given the function name)? With the current logic, I assume one could also send us a presigned JS file which may make use of Remote Code Execution? 🤔 Or do we safe-parse that within the frontend?

File: web/src/server/api/routers/utilities.ts [hunk 3]
```diff
@@ -81,17 +82,102 @@ const isValidAndSecureUrl = async (urlString: string): Promise<boolean> => {
...
+    } else {
+      logger.info("Unknown error:", error);
+    }
     return false;
   }
 };
 
 const isValidImageUrl = async (url: string): Promise<boolean> => {
   try {
+    if (await isValidPresignedS3Url(url)) {
```
Comment (Steffen911) [inline web/src/server/api/routers/utilities.ts:172]: Could you add a 1-2 sentence comment on why S3 presigned URLs deserve special behaviour here? The head fetch + content type check seems sound and it's not obvious why they wouldn't work with a S3 presigned URL.

Comment (marliessophie) [reply web/src/server/api/routers/utilities.ts:172]: Will do, yes 🙌

File: web/src/server/api/routers/utilities.ts (hunk 2 above)
Comment (marliessophie) [reply web/src/server/api/routers/utilities.ts:174]: Good shout, it is validated in the front-end also, but if someone reuses this function def a requirement to validate the content type as image also server side, will double check

File: web/src/server/api/routers/utilities.ts (hunk 1 above)
Comment (marliessophie) [reply web/src/server/api/routers/utilities.ts:108]: Not yet, I'd need to check if our current general check would catch these or not. Given no one has reported lack of support for Azure, GCP (or else) yet, I don't think we need to look into this right now :)

**Review by Steffen911 — APPROVED** [review]

**Pull request merged**